File: DemoApp/DemoApp/MainViewController.m
//
//  ViewController.m
//  DemoApp
//
//  Created by Shimi Sheetrit on 2/1/16.
//  Copyright © 2016 Matomy Media Group Ltd. All rights reserved.
//

#import "MainViewController.h"
#import "CollectionViewCell.h"
#import "SettingsViewController.h"
#import "NativeAdViewController.h"

#define ADS_TYPE_NUM 4
#define AD_REFRESH 0


#define MOBFOX_HASH_BANNER @"fe96717d9875b9da4339ea5367eff1ec" // @"9c09794cc85e7c9f9205e7a26f03234c"
#define MOBFOX_HASH_INTER @"267d72ac3f77a3f447b32cf7ebf20673"  // @"145849979b4c7a12916c7f06d25b75e3"
#define MOBFOX_HASH_NATIVE @"80187188f458cfde788d961b6882fd53" // @"4c3ea57788c5858881dc42cfafe8c0ab" 
#define MOBFOX_HASH_VIDEO @"651586294dac23e245f26789c4043aa9"



@interface MainViewController ()

@property (strong, nonatomic) MobFoxAd *mobfoxAd;
@property (strong, nonatomic) MobFoxInterstitialAd *mobfoxInterAd;
@property (strong, nonatomic) MobFoxNativeAd* mobfoxNativeAd;
@property (strong, nonatomic) MobFoxAd *mobfoxVideoAd;
@property (strong, nonatomic) NSURL *clickURL;
@property (strong, nonatomic) NSString *cellID;

@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;
@property (weak, nonatomic) IBOutlet UIView *nativeAdView;
@property (weak, nonatomic) IBOutlet UIView *innerNativeAdView;

@property (weak, nonatomic) IBOutlet UIImageView *nativeAdIcon;
@property (weak, nonatomic) IBOutlet UILabel *nativeAdTitle;
@property (weak, nonatomic) IBOutlet UILabel *nativeAdDescription;

@property (nonatomic) CGRect adVideoRect;



@end

@implementation MainViewController



- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    NSString * plistPath = [[NSBundle mainBundle] pathForResource:@"response.json" ofType:nil];
    NSLog(@"plistPath %@", plistPath);
    
    self.cellID = @"cellID";
    self.nativeAdView.hidden = YES;
    
    UITapGestureRecognizer *recognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleGesture:)];
    [self.innerNativeAdView addGestureRecognizer:recognizer];
    
    // Oreintation dependent in iOS 8 and later.
    float screenWidth = [UIScreen mainScreen].bounds.size.width;
    float screenHeight = [UIScreen mainScreen].bounds.size.height;
    float bannerWidth = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad ? 728.0 : 320.0;
    float bannerHeight = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad ? 90.0 : 50.0;
    float videoWidth = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad ? 500.0 : 300.0;
    float videoHeight = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad ? 450.0 : 250.0;

    
    /*** Banner ***/
    CGRect adRect = CGRectMake((screenWidth-bannerWidth)/2, screenHeight-bannerHeight, bannerWidth, bannerHeight);
    self.mobfoxAd = [[MobFoxAd alloc] init:MOBFOX_HASH_BANNER withFrame:adRect];
    self.mobfoxAd.delegate = self;
    self.mobfoxAd.refresh = [NSNumber numberWithInt:AD_REFRESH];
    //self.mobfoxAd.type = @"video";
    //self.mobfoxAd.v_dur_min = [NSNumber numberWithInt:10];
    //self.mobfoxAd.v_dur_max = [NSNumber numberWithInt:2000];
    [self.view addSubview:self.mobfoxAd];
    
    /*** Interstitial ***/
    self.mobfoxInterAd = [[MobFoxInterstitialAd alloc] init:MOBFOX_HASH_INTER withRootViewController:self];
    //self.mobfoxInterAd.ad.type = @"video";
    self.mobfoxInterAd.delegate = self;
    
    /*** Native ***/
    self.mobfoxNativeAd = [[MobFoxNativeAd alloc] init:MOBFOX_HASH_NATIVE];
    self.mobfoxNativeAd.delegate = self;
    
    /*** Video ***/
    float videoTopMargin = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad ? 200.0 : 80.0;
    self.adVideoRect = CGRectMake((screenWidth - videoWidth)/2, self.collectionView.frame.size.height + videoTopMargin, videoWidth, videoHeight);
    [self initVideoAd];
    
    
    ////////////////
    /*MPServerAdPositioning *positioning = [[MPServerAdPositioning alloc] init];
    
    self.placer = [MPCollectionViewAdPlacer placerWithCollectionView:self.collectionView
                                                      viewController:self
                                                       adPositioning:positioning
                                             defaultAdRenderingClass:[YourNativeAdCell class]];
    [self.placer loadAdsForAdUnitID:self.adUnitID];*/
    
    //MPNativeAd *native  = [MPNativeAd alloc] initWithAdAdapter:<#(id<MPNativeAdAdapter>)#>
    /*
    MPStaticNativeAdRendererSettings *settings = [[MPStaticNativeAdRendererSettings alloc] init];
    settings.renderingViewClass = [self.view class];
    MPNativeAdRendererConfiguration *config = [MPStaticNativeAdRenderer rendererConfigurationWithRendererSettings:settings];
    MPNativeAdRequest *adRequest = [MPNativeAdRequest requestWithAdUnitIdentifier:@"ac0f139a2d9544fface76d06e27bc02a" rendererConfigurations:@[config]];

    MPNativeAdRequestTargeting *targeting = [MPNativeAdRequestTargeting targeting];
    targeting.desiredAssets = [NSSet setWithObjects:kAdTitleKey, kAdTextKey, kAdCTATextKey, kAdIconImageKey, kAdMainImageKey, kAdStarRatingKey, nil]; //The constants correspond to the 6 elements of MoPub native ads
    
    [adRequest startWithCompletionHandler:^(MPNativeAdRequest *request, MPNativeAd *response, NSError *error) {
        if (error) {
            // Handle error.
        } else {
            
            NSLog(@"response: %@", response);
     
            //self.nativeAd = response;
            //self.nativeAd.delegate = self;
            //UIView *nativeAdView = [response retrieveAdViewWithError:nil];
            //nativeAdView.frame = self.yourNativeAdViewContainer.bounds;
            //[self.yourNativeAdViewContainer addSubview:nativeAdView];
     
        }
    }];*/


    
    
}

- (void)viewWillDisappear:(BOOL)animated {
    
    NSLog(@"viewWillDisappear");
    
    [super viewWillDisappear:animated];
    [self removeVideoAd];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark Collection View Data Source

- (NSInteger)numberOfSectionsInCollectionView: (UICollectionView *)collectionView {
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)view numberOfItemsInSection:(NSInteger)section {
    return ADS_TYPE_NUM;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)cv cellForItemAtIndexPath:(NSIndexPath *)indexPath;
{

    CollectionViewCell *cell = [cv dequeueReusableCellWithReuseIdentifier:self.cellID forIndexPath:indexPath];
    cell.title.text = [self adTitle:indexPath];
    cell.image.image = [self adImage:indexPath];
    
    if (cell.selected) {
        cell.backgroundColor = [UIColor lightGrayColor];
    } else {
        cell.backgroundColor = [UIColor whiteColor]; // Default color
    }
    
    return cell;
}

#pragma mark Collection View Delegate

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    
    
    UICollectionViewCell* cell = [collectionView  cellForItemAtIndexPath:indexPath];
    cell.backgroundColor = [UIColor lightGrayColor];
    
    
    switch (indexPath.item) {
        case 0:
            [self hideAds:indexPath];
            [self removeVideoAd];
            self.mobfoxAd.invh = self.invh.length > 0 ? self.invh: MOBFOX_HASH_BANNER;
            [self resumeAdRefresh];
            break;
            
        case 1:
            
            [self hideAds:indexPath];
            [self pauseAdRefresh];
            [self removeVideoAd];
            self.mobfoxInterAd.ad.invh = self.invh.length > 0 ? self.invh: MOBFOX_HASH_INTER;
            [self.mobfoxInterAd loadAd];
            break;
            
        case 2:
            [self hideAds:indexPath];
            [self pauseAdRefresh];
            [self removeVideoAd];
            self.mobfoxNativeAd.invh = self.invh.length > 0 ? self.invh: MOBFOX_HASH_NATIVE;
            [self.mobfoxNativeAd loadAd];
            //[self presentViewController];
            
            break;
            
        case 3:
            
            [self hideAds:indexPath];
            [self pauseAdRefresh];
            [self removeVideoAd];
            [self initVideoAd];
            self.mobfoxVideoAd.invh = self.invh.length > 0 ? self.invh: MOBFOX_HASH_VIDEO;
            [self.mobfoxVideoAd loadAd];
            break;
            
        default:
            break;
    }
    
}

- (void)collectionView:(UICollectionView *)collectionView didDeselectItemAtIndexPath:(NSIndexPath *)indexPath{
    UICollectionViewCell* cell = [collectionView  cellForItemAtIndexPath:indexPath];
    cell.backgroundColor = [UIColor whiteColor];
}



#pragma mark MobFox Ad Delegate

//called when ad is displayed
- (void)MobFoxAdDidLoad:(MobFoxAd *)banner {
    
    NSLog(@"MobFoxAdDidLoad:");
}

//called when an ad cannot be displayed
- (void)MobFoxAdDidFailToReceiveAdWithError:(NSError *)error {
    
    NSLog(@"MobFoxAdDidFailToReceiveAdWithError: %@", [error description]);
}

//called when ad is closed/skipped
- (void)MobFoxAdClosed {
    NSLog(@"MobFoxAdClosed:");

}

//called when ad is clicked
- (void)MobFoxAdClicked {
    NSLog(@"MobFoxAdClicked:");

}

#pragma mark MobFox Interstitial Ad Delegate

//best to show after delegate informs an ad was loaded
- (void)MobFoxInterstitialAdDidLoad:(MobFoxInterstitialAd *)interstitial {
    
    NSLog(@"MobFoxInterstitialAdDidLoad:");
    
    if(self.mobfoxInterAd.ready){
        [self.mobfoxInterAd show];
    }

}

- (void)MobFoxInterstitialAdDidFailToReceiveAdWithError:(NSError *)error {
    
    NSLog(@"MobFoxInterstitialAdDidFailToReceiveAdWithError: %@", [error description]);
    
}

- (void)MobFoxInterstitialAdClosed {
    
    NSLog(@"MobFoxInterstitialAdClosed");
    
}

- (void)MobFoxInterstitialAdClicked {
    
    NSLog(@"MobFoxInterstitialAdClicked");
    
}

- (void)MobFoxInterstitialAdFinished {
    
    NSLog(@"MobFoxInterstitialAdFinished");
    
}

#pragma mark MobFox Native Ad Delegate

//called when ad response is returned
- (void)MobFoxNativeAdDidLoad:(MobFoxNativeAd *)ad withAdData:(MobFoxNativeData *)adData {
    
    self.nativeAdIcon.image = [UIImage imageWithData:[NSData dataWithContentsOfURL:adData.icon.url]];
    self.nativeAdTitle.text = adData.assetHeadline;
    self.nativeAdDescription.text = adData.assetDescription;
    self.clickURL = [adData.clickURL absoluteURL];
    
    for (MobFoxNativeTracker *tracker in adData.trackersArray) {
        
        // Fire tracking pixel
        UIWebView* wv = [[UIWebView alloc] initWithFrame:CGRectZero];
        NSString* userAgent = [wv stringByEvaluatingJavaScriptFromString:@"navigator.userAgent"];
        NSURLSessionConfiguration* conf = [NSURLSessionConfiguration defaultSessionConfiguration];
        [conf.HTTPAdditionalHeaders setValue:userAgent forKey:@"User-Agent"];
        NSURLSession* session = [NSURLSession sessionWithConfiguration:conf];
        NSURLSessionDataTask* task = [session dataTaskWithURL:tracker.url completionHandler:
                                      ^(NSData *data,NSURLResponse *response, NSError *error){
                                          
                                          if(error) NSLog(@"err %@",[error description]);

                                      }];
        [task resume];
        
    }
    
}

//called when ad response cannot be returned
- (void)MobFoxNativeAdDidFailToReceiveAdWithError:(NSError *)error {
    
    NSLog(@"MobFoxNativeAdDidFailToReceiveAdWithError: %@", [error description]);
    
}

#pragma mark Private Methods

- (void)hideAds:(NSIndexPath *)indexPath {
    
    switch (indexPath.item) {
        case 0:
            self.mobfoxAd.hidden= NO;
            self.mobfoxInterAd.ad.hidden = YES;
            self.nativeAdView.hidden = YES;
            self.mobfoxVideoAd.hidden = YES;
            
            break;
            
        case 1:
            self.mobfoxAd.hidden= YES;
            self.mobfoxInterAd.ad.hidden = NO;
            self.nativeAdView.hidden = YES;
            self.mobfoxVideoAd.hidden = YES;
            
            break;
            
        case 2:
            self.mobfoxAd.hidden= YES;
            self.mobfoxInterAd.ad.hidden = YES;
            self.nativeAdView.hidden = NO;
            self.mobfoxVideoAd.hidden = YES;
            
            break;
            
        case 3:
            self.mobfoxAd.hidden= YES;
            self.mobfoxInterAd.ad.hidden = YES;
            self.nativeAdView.hidden = YES;
            self.mobfoxVideoAd.hidden = NO;
            
            break;
            
        default:
            break;
    }
    
}

- (NSString *)adTitle:(NSIndexPath *)indexPath {
    
    switch (indexPath.item) {
        case 0:
            return @"Banner";
            break;
        case 1:
            return @"Interstitial";
            break;
        case 2:
            return @"Native";
            break;
        case 3:
            return @"Video";
            break;
            
        default:
            return @"";
            break;
    }
}

- (UIImage *)adImage:(NSIndexPath *)indexPath {
    
    switch (indexPath.item) {
        case 0:
            return [UIImage imageNamed:@"test_banner.png"];
            break;
        case 1:
            return [UIImage imageNamed:@"test_interstitial.png"];
            break;
        case 2:
            return [UIImage imageNamed:@"test_native.png"];
            break;
        case 3:
            return [UIImage imageNamed:@"test_video.png"];
            break;
            
        default:
            return nil;
            break;
    }
}

- (void)handleGesture:(UIGestureRecognizer *)gestureRecognizer {
    
    [[UIApplication sharedApplication] openURL:self.clickURL];
    
}

- (void)initVideoAd {
    
    if (self.mobfoxVideoAd == nil) {
        
        self.mobfoxVideoAd = [[MobFoxAd alloc] init:MOBFOX_HASH_VIDEO withFrame:self.adVideoRect];
        self.mobfoxVideoAd.delegate = self;
        self.mobfoxVideoAd.type = @"video";
        
        [self.view addSubview:self.mobfoxVideoAd];
    }
    
}

- (void)pauseAdRefresh {
    
    if (AD_REFRESH > 0) {
        
        self.mobfoxAd.refresh = [NSNumber numberWithInt:0];
        [self.mobfoxAd loadAd];
        [self.mobfoxAd removeFromSuperview];
    }
}

- (void)resumeAdRefresh {
    
    if (AD_REFRESH > 0) {
        
        [self.view addSubview:self.mobfoxAd];
        self.mobfoxAd.refresh = [NSNumber numberWithInt:AD_REFRESH];
        [self.mobfoxAd loadAd];
        
    } else {
        
        self.mobfoxAd.refresh = [NSNumber numberWithInt:AD_REFRESH];
        [self.mobfoxAd loadAd];
    }
}

- (void)removeVideoAd {
    
    [self.mobfoxVideoAd stopPlayback];
    self.mobfoxVideoAd = nil;
    
}

- (void)presentViewController {
    
    NativeAdViewController *nativeVC = [[NativeAdViewController alloc] init];
    [self presentViewController:nativeVC animated:YES completion:nil];
    
}


@end

File: DemoAppSwift/DemoAppSwift/MainViewController.swift
//
//  ViewController.swift
//  DemoAppSwift
//
//  Created by Shimi Sheetrit on 2/16/16.
//  Copyright © 2016 Matomy Media Group Ltd. All rights reserved.
//

import UIKit
import MobFoxSDKCore
import GoogleMobileAds


let AdsTypeNum = 4
let adRefresh = 10

struct InventoryHash {
    
    static let MobFoxHashBanner = "fe96717d9875b9da4339ea5367eff1ec"
    static let MobFoxHashInter = "267d72ac3f77a3f447b32cf7ebf20673"
    static let MobFoxHashNative = "80187188f458cfde788d961b6882fd53"
    static let MobFoxHashVideo = "651586294dac23e245f26789c4043aa9"
}

class MainViewController: UIViewController, UICollectionViewDelegate, UICollectionViewDataSource, MobFoxAdDelegate, MobFoxInterstitialAdDelegate, MobFoxNativeAdDelegate {
    
    @IBOutlet weak var collectionView: UICollectionView!
    @IBOutlet weak var nativeAdView: UIView!
    @IBOutlet weak var innerNativeAdView: UIView!
    @IBOutlet weak var nativeAdTitle: UILabel!
    @IBOutlet weak var nativeAdDescription: UILabel!
    @IBOutlet weak var nativeAdIcon: UIImageView!
    
    internal var invh: String?
    private let cellID = "cellID"
    private var clickURL: URL?
    private var adVideoRect: CGRect!
    
    // MobFox.
    private var mobfoxAd: MobFoxAd!
    private var mobfoxInterAd: MobFoxInterstitialAd!
    private var mobfoxNativeAd: MobFoxNativeAd!
    private var mobfoxVideoAd: MobFoxAd!
    
    // AdMob.
    private var bannerView: GADBannerView!
    

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.

        
        self.nativeAdView.isHidden = true
        
        let recognizer = UITapGestureRecognizer.init(target: self, action: "handleGesture:")
        self.innerNativeAdView.addGestureRecognizer(recognizer)
        
        // Oreintation dependent in iOS 8 and later.
        let screenWidth = UIScreen.main().bounds.size.width
        let screenHeight = UIScreen.main().bounds.size.height
        let bannerWidth = CGFloat(UIDevice.current().userInterfaceIdiom == UIUserInterfaceIdiom.pad ? 728.0 : 320.0)
        let bannerHeight = CGFloat(UIDevice.current().userInterfaceIdiom == UIUserInterfaceIdiom.pad ? 90.0 : 50.0)
        let videoWidth = CGFloat(UIDevice.current().userInterfaceIdiom == UIUserInterfaceIdiom.pad ? 500.0 : 300.0)
        let videoHeight = CGFloat(UIDevice.current().userInterfaceIdiom == UIUserInterfaceIdiom.pad ? 450.0 : 250.0)

        /*** Banner ***/
        let rect = CGRect(origin: CGPoint(x: (screenWidth-bannerWidth)/2, y: screenHeight-bannerHeight), size: CGSize(width: bannerWidth, height: bannerHeight) )
        mobfoxAd = MobFoxAd(InventoryHash.MobFoxHashBanner, withFrame: rect)
        mobfoxAd.delegate = self
        mobfoxAd.refresh = adRefresh
        self.view.addSubview(mobfoxAd)
        
        /*** Interstitial ***/
        mobfoxInterAd = MobFoxInterstitialAd(InventoryHash.MobFoxHashInter, withRootViewController:self)
        mobfoxInterAd.delegate = self

        /*** Native ***/
        mobfoxNativeAd = MobFoxNativeAd(InventoryHash.MobFoxHashNative)
        mobfoxNativeAd.delegate = self
        
        /*** Video ***/
        let videoTopMargin = CGFloat(UIDevice.current().userInterfaceIdiom == UIUserInterfaceIdiom.pad ? 200.0 : 80.0)
        self.adVideoRect = CGRect(x: (screenWidth - videoWidth)/2, y: self.collectionView.frame.size.height + videoTopMargin, width: videoWidth, height: videoHeight)
        self.initVideoAd()
        
        /*** AdMob ***/
        bannerView = GADBannerView.init(adSize: kGADAdSizeBanner, origin: CGPoint(x: 0, y: 0))
        bannerView.adUnitID = "ca-app-pub-6224828323195096/5240875564​" // "ca-app-pub-3940256099942544/2934735716"
        bannerView.frame = CGRect(x: 0, y: 0, width: 320, height: 50)
        bannerView.rootViewController = self
        let request = GADRequest()
        //request.testDevices = ["221e6c438e8184e0556942ea14bb214b"] // kGADSimulatorID
        bannerView.load(request)
        self.view.addSubview(bannerView)
        
     
        
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        
        print("viewWillDisappear")
        
        super.viewWillDisappear(animated)
        self.removeVideoAd()

    }
    
    //MARK: MobFox Ad Delegate
     func mobFoxAdDidLoad(_ banner: MobFoxAd!) {
        
        print("MobFoxAdDidLoad")
        
    }
    
    func mobFoxAdDidFailToReceiveAdWithError(_ error: NSError!) {
        
        print("MobFoxAdDidFailToReceiveAdWithError: ", error.description)
        
    }
    
    func mobFoxAdClosed() {
        
        print("MobFoxAdClosed")

    }
    
    func mobFoxAdClicked() {
        
        print("MobFoxAdClicked")

    }
    
    //MARK: MobFox Ad Interstitial Delegate
    func mobFoxInterstitialAdDidLoad(_ interstitial: MobFoxInterstitialAd!) {
        
        print("MobFoxInterstitialAdDidLoad")
        
        if(self.mobfoxInterAd.ready){
            self.mobfoxInterAd.show()
        }

    }
    
    func mobFoxInterstitialAdDidFailToReceiveAdWithError(_ error :NSError!) {
    
        print("MobFoxInterstitialAdDidFailToReceiveAdWithError: \(error.description)");
    
    }
    
    func mobFoxInterstitialAdClosed() {
    
    print("MobFoxInterstitialAdClosed")
    
    }
    
    func mobFoxInterstitialAdClicked() {
    
    print("MobFoxInterstitialAdClicked")
    
    }
    
    func mobFoxInterstitialAdFinished() {
    
    print("MobFoxInterstitialAdFinished")
    
    }
    
    //MARK: MobFox Ad Native Delegate
    func mobFoxNativeAdDidLoad(_ ad: MobFoxNativeAd!, withAdData adData: MobFoxNativeData!) {
        
        self.nativeAdIcon.image = UIImage(data: try! Data(contentsOf: adData.icon.url))
        self.nativeAdTitle.text = adData.assetHeadline
        self.nativeAdDescription.text = adData.assetDescription
        self.clickURL = adData.clickURL.absoluteURL
                
    }
    
    func mobFoxNativeAdDidFailToReceiveAdWithError(_ error: NSError!) {
        
        print("MobFoxNativeAdDidFailToReceiveAdWithError")

    }
    
    //MARK: UICollection View Delegate
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        
        print("collectionView:didSelectItemAtIndexPath")
        
        let cell = collectionView.cellForItem(at: indexPath)
        if cell != nil {
            cell!.backgroundColor = UIColor.lightGray()
        }
        
        switch (indexPath as NSIndexPath).item {
        case 0:
            self.hideAds(indexPath)
            self.removeVideoAd()
            if self.invh != nil {self.mobfoxAd.invh = self.invh!.characters.count > 0 ? self.invh! : InventoryHash.MobFoxHashBanner}
            self.resumeAdRefresh()
            
        case 1:
            self.hideAds(indexPath)
            self.pauseAdRefresh()
            self.removeVideoAd()
            if self.invh != nil {self.mobfoxInterAd.ad.invh = self.invh!.characters.count > 0  ? self.invh! : InventoryHash.MobFoxHashInter}
            self.mobfoxInterAd.loadAd()

        case 2:
            self.hideAds(indexPath)
            self.pauseAdRefresh()
            self.removeVideoAd()
            if self.invh != nil {self.mobfoxNativeAd.invh = self.invh!.characters.count > 0  ? self.invh! : InventoryHash.MobFoxHashNative}
            self.mobfoxNativeAd.load()

        case 3:
            self.hideAds(indexPath)
            self.pauseAdRefresh()
            self.removeVideoAd()
            self.initVideoAd()
            if self.invh != nil {self.mobfoxVideoAd.invh = self.invh!.characters.count > 0  ? self.invh! : InventoryHash.MobFoxHashVideo}
            self.mobfoxVideoAd.load()

            
        default:
             break
        }
        
    }
    
    func collectionView(_ collectionView: UICollectionView, didDeselectItemAt indexPath: IndexPath) {
        
        let cell = collectionView.cellForItem(at: indexPath)
        if cell != nil {
            cell?.backgroundColor = UIColor.white()
        }
        
    }
    
    //MARK: UICollection View Data Source
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: self.cellID , for: indexPath) as! CollectionViewCell
        
        cell.title.text = self.adTitle(indexPath)
        cell.image.image = self.adImage(indexPath)

        if cell.isSelected {
            cell.backgroundColor = UIColor.lightGray()
        } else {
            cell.backgroundColor = UIColor.white() // Default color
        }
        
        return cell
        
    }
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        
        return 1
        
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        
        return 4
    }
    
    //MARK: Private Functions
    private func adTitle(_ indexPath: IndexPath) ->String {
        
        switch (indexPath as NSIndexPath).item {
        case 0:
            return "Banner"
        case 1:
            return "Interstitial"
        case 2:
            return "Native"
        case 3:
            return "Video"
            
        default:
            return ""
        }
        
    }
    
    private func adImage(_ indexPath: IndexPath) ->UIImage {
        
        switch (indexPath as NSIndexPath).item {
        case 0:
            return UIImage(named:"test_banner.png")!
        case 1:
            return UIImage(named:"test_interstitial.png")!
        case 2:
            return UIImage(named:"test_native.png")!
        case 3:
            return UIImage(named:"test_video.png")!
            
        default:
            return UIImage()
        }
        
    }
    
    private func hideAds(_ indexPath: IndexPath) {
        
        switch (indexPath as NSIndexPath).item {
            
        case 0:
            self.mobfoxAd.isHidden = false
            self.mobfoxInterAd.ad.isHidden = true
            self.nativeAdView.isHidden = true
            if self.mobfoxVideoAd != nil {self.mobfoxVideoAd.isHidden = true}
            
        case 1:
            self.mobfoxAd.isHidden = true
            self.mobfoxInterAd.ad.isHidden = false
            self.nativeAdView.isHidden = true
            if self.mobfoxVideoAd != nil {self.mobfoxVideoAd.isHidden = true}
            
        case 2:
            self.mobfoxAd.isHidden = true
            self.mobfoxInterAd.ad.isHidden = true
            self.nativeAdView.isHidden = false
            if self.mobfoxVideoAd != nil {self.mobfoxVideoAd.isHidden = true}
            
        case 3:
            self.mobfoxAd.isHidden = true
            self.mobfoxInterAd.ad.isHidden = true
            self.nativeAdView.isHidden = true
            if self.mobfoxVideoAd != nil {self.mobfoxVideoAd.isHidden = true}
            
        default:
            break
            
        }
        
    }
        
    private func initVideoAd() {
        
        if (self.mobfoxVideoAd == nil) {
            
            self.mobfoxVideoAd = MobFoxAd(InventoryHash.MobFoxHashVideo, withFrame: self.adVideoRect)
            self.mobfoxVideoAd.delegate = self
            self.mobfoxVideoAd.type = "video"
            self.view.addSubview(self.mobfoxVideoAd)
        }
    }
    
    private func removeVideoAd () {
        
        if self.mobfoxVideoAd != nil {
        
        //self.mobfoxVideoAd.stopPlayback()
        self.mobfoxAd.removeFromSuperview()
        self.mobfoxVideoAd = nil
        
        }
        
    }
    
    private func pauseAdRefresh (){
        
        if(adRefresh > 0) {
            
            self.mobfoxAd.refresh = NSNumber.init(value: 0)
            self.mobfoxAd.load()
            self.mobfoxAd.removeFromSuperview()
        }
    }
    
    private func resumeAdRefresh () {
        
        if(adRefresh > 0) {
            
            self.view.addSubview(self.mobfoxAd)
            self.mobfoxAd.refresh = NSNumber.init(value: adRefresh)
            self.mobfoxAd.load()
            
        } else {
            
            self.mobfoxAd.refresh = NSNumber.init(value: adRefresh)
            self.mobfoxAd.load()
        }
    }
    
    private func handleGesture( _ gestureRecognizer: UIGestureRecognizer) {
        
        if (self.clickURL != nil) {
            UIApplication.shared().openURL(self.clickURL!)
        }
    }
    
}

